compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -649,29 +649,334 @@ abstract class GenJSCode extends plugins.PluginComponent
         constructorTrees: List[DefDef]): js.Tree = {
       implicit val pos = classSym.pos
 
-      val (primaryCtorTree :: Nil, secondaryCtorTrees) =
-        constructorTrees.partition(_.symbol.isPrimaryConstructor)
-
       // Implementation restriction
-      val sym = primaryCtorTree.symbol
+      val syms = constructorTrees.map(_.symbol)
       val hasBadParam = enteringPhase(currentRun.uncurryPhase) {
-        sym.paramss.flatten.exists(p => p.hasDefault || isRepeated(p))
+        syms.exists(_.paramss.flatten.exists(p => p.hasDefault))
       }
       if (hasBadParam) {
         reporter.error(pos,
             "Implementation restriction: the constructor of a " +
-            "Scala.js-defined JS classes cannot have default parameters nor " +
-            "repeated parameters.")
+            "Scala.js-defined JS classes cannot have default parameters.")
       }
 
-      // Implementation restriction
-      for (tree <- secondaryCtorTrees) {
-        reporter.error(tree.pos,
-            "Implementation restriction: Scala.js-defined JS classes cannot " +
-            "have secondary constructors")
+      withNewLocalNameScope {
+        val ctors: List[js.MethodDef] = constructorTrees.flatMap { tree =>
+          genMethodWithCurrentLocalNameScope(tree)
+        }
+
+        val dispatch =
+          genJSConstructorExport(constructorTrees.map(_.symbol))
+        val js.MethodDef(_, dispatchName, dispatchArgs, dispatchResultType,
+            dispatchResolution) = dispatch
+
+        val isConstructor = ctors.map(_.name.name).toSet
+
+        val constructorTree = mkConstructorTree(ctors, isConstructor)
+
+        val overloadIdent = freshLocalIdent("overload")
+
+        // Section containing the overload resolution and casts of parameters
+        val overloadSelection = mkOverloadSelection(constructorTree,
+            overloadIdent, dispatchResolution, isConstructor)
+
+        /* Section containing all the code executed before the call to `this`
+         * for every secondary constructor.
+         */
+        val prePrimaryCtorBody =
+          constructorTree.mkPreSuperCall(overloadIdent, isConstructor)
+
+        val primaryCtorBody = constructorTree.rootBody
+
+        /* Section containing all the code executed after the call to this for
+         * every secondary constructor.
+         */
+        val postPrimaryCtorBody =
+          constructorTree.mkPostSuperCall(overloadIdent, isConstructor)
+
+        val newBody = js.Block(overloadSelection ::: prePrimaryCtorBody ::
+            primaryCtorBody :: postPrimaryCtorBody :: Nil)
+
+        js.MethodDef(static = false, dispatchName, dispatchArgs, jstpe.NoType,
+            newBody)(dispatch.optimizerHints, dispatch.hash)
+      }
+    }
+
+    private class ConstructorTree(overrideNum: Int, method: js.MethodDef,
+        subConstructors: List[ConstructorTree]) {
```

Comment:
This class seems to conflate two things: the tree data structure; and the construction methods. Apparently, all construction methods must only be called on the top-level node of this tree. Consider separating the construction methods into another class `JSConstructorBuilder(root: ConstructorTree)`, and keep `ConstructorTree` as a pure data structure.